Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb

```diff
@@ -117,15 +140,63 @@ def process(queue)
           # Work-around: After returned by get_blob, the etag will contains quotes.
           new_etag = blob.properties[:etag]
           # ~ Work-around
-          blob, content = @azure_blob.get_blob(@container, blob_name, {:start_range=>start_index} )
-          
-          @codec.decode(content) do |event|
-            decorate(event)
-            queue << event
-          end # decode
+
+          blob, header = @azure_blob.get_blob(@container, blob_name, {:end_range => (@file_head_bytes-1) }) if header.nil? unless @file_head_bytes.nil? or @file_head_bytes <= 0
+
+          if start_index == 0
+            # Skip the header since it is already read.
+            start_index = @file_head_bytes
+          else
+            # Adjust the offset when it is other than first time, then read till the end of the file, including the tail.
+            start_index = start_index - @file_tail_bytes
+            start_index = 0 if start_index < 0
+          end
+
+          blob, content = @azure_blob.get_blob(@container, blob_name, {:start_range => start_index} )
+
+          # content will be used to calculate the new offset. Create a new variable for processed content.
+          processed_content = content
+
+          is_json_codec = @codec.is_a? LogStash::Codecs::JSON
+          if (is_json_codec)
+            skip = processed_content.index '{'
+            processed_content = processed_content[skip..-1] unless skip.nil?
```

Comment:
start_index may start somewhere in the middle of the file. What's the reason for skipping the first '{' that you see in the middle of the file?

Also, I don't know if line 163 will cause a copy of the entire array. Since the blob is potentially large, it may be good to find ways to avoid this copying.

Reply:
@brahmnes,
1. You read it correct that it might start in the middle except that it skips any characters before the first `{`. 
2. I can look into the copy content issue. That is a valid consideration for hyper-scale scenario, although by benchmark, I don't see it blown up memory for logs at the size of 300MB.

Reply:
Got your point about skipping to the first '{'.
For #2, the memory is a concern, but the bigger concern is speed. If this does get up to several hundreds of MB, and your agent is trying to process many of these blobs, then it can make a difference.